Class Tweakers: Enum Extensions (#567)
FabricMC/fabric-docs: develop/class-tweakers/enum-extension.md

For example, the class tweaker entries for the constants we added in the [mixin section](#creating-the-mixin) would respectively be:

<<< @/reference/latest/src/main/resources/example-mod.classtweaker#enum-extension-no-impls-example-entry

and

<<< @/reference/latest/src/main/resources/example-mod.classtweaker#enum-extension-ctor-impls-example-entry

## Applying Changes {#applying-changes}

You'll have to refresh your Gradle project by [regenerating sources](../getting-started/generating-sources) before you can see your added enum entries in the decompiled source.
If modifications do not appear, you can try [validating](../class-tweakers/index#validating-the-file) the file and checking if any errors appear.

::: info

You will not see [fields, methods or other elements](#additional-fields-and-methods) in the decompiled source code. That's because those are handled by the mixin, and are only applied at runtime.

:::

You can now use the enum constant in your code:

<<< @/reference/latest/src/main/java/com/example/docs/enum_extension/ExampleModEnumExtension.java#enum-extension-added-constant-usage-example

> If you are only adding it with a mixin and it is not in the decompiled source, get it by calling `valueOf`:

Reviewer: ```suggestion
If you are only adding it with a mixin and it is not in the decompiled source, obtain it by calling `valueOf`:
```

Nit: "get it" is correct, but it can also mean "understand it" so I avoided it.

<<< @/reference/latest/src/main/java/com/example/docs/enum_extension/ExampleModEnumExtension.java#enum-extension-added-constant-no-ct-usage-example

## Pitfalls {#pitfalls}

Enum extension by itself makes no guarantees that your added entry will not break anything. You should find the modified enum's usages to try and

> prevent issues where possible. If an issue that would cause a crash arises and it cannot be fixed, it may be best to not use enum extension at all.

Reviewer: ```suggestion
Enum extension cannot guarantee that entries you add will not break anything.

It's your responsibility to go through the target enum's usages, and try to
prevent issues where possible. If you are unable to solve some, and crashes arise, it might be best to not use enum extension at all.
```

> This section goes over patterns to look out for and to avoid in your own code, but it is in no way exhaustive.

Reviewer: ```suggestion
This section goes over some patterns to watch out for and avoid when extending enums, but it is not exhaustive.
```

### Switch Expressions {#switch-expressions}

Switches are often used to handle enum constants, this can lead to a crash if a switch expression has no `default` case to handle

> entries added by mods. For example, if we had the following switch expression:

Reviewer: ```suggestion
Switch statements are often used to handle enum constants. Because of this, a crash can happen if a switch expression does not handle
entries added by other mods. For example, say we have the following switch expression:
```

<<< @/reference/latest/src/main/java/com/example/docs/enum_extension/ExampleModEnumExtension.java#enum-extension-problematic-switch-expr-example

> Even though we handle all the values that are in the Vanilla enum and our own, this would throw for any other mod's added entry.

Reviewer: ```suggestion
Notice how there is no `default` clause. Even though we handled all the values in the Vanilla enum, and our own, this would throw if another mod adds a different entry.
```

If the expression is from a mod, you should consider raising an issue to the developers to find a more compatible option.
If the expression is in a Vanilla method or you are unable to get the issue fixed by the mod developers, a mixin is likely needed.

> There is no one way for mixins to modify switch expressions to avoid throws, and the approach should be made on a case-by-case basis.

Reviewer: I flipped the order of the paragraphs:

```suggestion
How can you prevent this? There is no single universal way to avoid crashes - your approach should be adapted on a case-by-case basis. In general, though:

- If the `switch` expression is in a vanilla method, you can use a mixin to edit it
- If the `switch` expression comes from a mod, you should try getting in contact with the developers, to work together a compatible approach. Otherwise, you might have to create a mixin on the other mod.
```

Author: You made a typo on `contact` on line 120, you typed `cnotact` instead.

Reviewer: You dare use my own spell against me, Potter..

(jokes aside, good catch!)

### Serialized Enums {#serialized-enums}

Certain enums may have their entries serialized automatically, such as the static `Variants` enum in the `Axolotl` class. Adding an entry to these enums
would make your entry serialized under Minecraft's namespace by the codec, on some versions having it be serialized based on a numerical ID.

> It is best to avoid extending enums whose entries are serialized like this entirely, and if possible looking for an API.

Reviewer: ```suggestion
Certain enums' entries get serialized automatically. An example is the `Variants` enum in the `Axolotl` class.

Extending these enums
would serialize your custom entry under Minecraft's namespace, and on some versions that might happen based on a numerical ID.
This is not great because it can affect the indices of all other entries.

It is best to avoid extending enums entirely if their entries are serialized like this. Instead, you might want to look for an API, if available.
```

correct me

Additionally, class tweaker file headers must specify `v2` as the version to use enum extensions.

:::

## Creating the Mixin {#creating-the-mixin}

Before creating the mixin class, make sure that Loader 0.19.0 or above is an explicit dependency
in your `fabric.mod.json` file:

```json:no-line-numbers
...
"depends": {
  ...
  "fabricloader": ">=0.19.0"
  ...
}
...
```

Even if you are using the correct Loader version as a Gradle dependency, you must explicitly depend on at least version 0.19.0 in
order to opt into this Mixin feature.

To make an enum extension, create an `enum` in your mixin package, annotate it with `@Mixin`, and add your constants to it as if they were
part of the targeted enum class. For example, let's add a new entry to `RecipeBookType`:

<<< @/reference/latest/src/main/java/com/example/docs/mixin/class_tweakers/RecipeBookTypeMixin.java#enum-extension-no-impls-example-mixin

::: warning IMPORTANT

You should always prefix the enum constants you add with your mod ID to ensure uniqueness. For these Docs, we'll use `EXAMPLE_MOD_`.

:::

### Additional Fields and Methods {#additional-fields-and-methods}

> If the targeted enum declares additional fields or method implementations, you'll have to shadow the relevant constructor.

Reviewer 2: Not really true, it can have fields and methods but a default constructor. The key points are that if you want to use a non-default constructor you have to shadow it, and separately if the enum has abstract methods you must shadow and implement those in your added entries.

Additionally, class tweaker file headers must specify `v2` as the version to use enum extensions.

:::

## Creating the Mixin {#creating-the-mixin}

Before creating the mixin class, make sure that Loader 0.19.0 or above is an explicit dependency
in your `fabric.mod.json` file:

```json:no-line-numbers
...
"depends": {
  ...
  "fabricloader": ">=0.19.0"
  ...
}
...
```

Even if you are using the correct Loader version as a Gradle dependency, you must explicitly depend on at least version 0.19.0 in
order to opt into this Mixin feature.

To make an enum extension, create an `enum` in your mixin package, annotate it with `@Mixin`, and add your constants to it as if they were
part of the targeted enum class. For example, let's add a new entry to `RecipeBookType`:

<<< @/reference/latest/src/main/java/com/example/docs/mixin/class_tweakers/RecipeBookTypeMixin.java#enum-extension-no-impls-example-mixin

::: warning IMPORTANT

You should always prefix the enum constants you add with your mod ID to ensure uniqueness. For these Docs, we'll use `EXAMPLE_MOD_`.

:::

### Specifying Fields {#specifying-fields}

> If the target enum has fields to initialize and no default, no-arg, constructor, you must shadow a target class constructor to

Reviewer 2: Nit: fields aren't relevant, only the presence of a no-arg constructor is

## Pitfalls {#pitfalls}

Enum extension cannot guarantee that entries you add will not break anything.

It's your responsibility to go through the target enum's usages, and try to
prevent issues where possible. If you are unable to solve some, and crashes arise, it might be best to not use enum extension at all.

This section goes over some patterns to watch out for and avoid when extending enums, but it is not exhaustive.

### Switch Expressions {#switch-expressions}

Switch statements are often used to handle enum constants. Because of this, a crash can happen if a switch expression does not handle
entries added by other mods. For example, say we have the following switch expression:

<<< @/reference/latest/src/main/java/com/example/docs/enum_extension/ExampleModEnumExtension.java#enum-extension-problematic-switch-expr-example

Notice how there is no `default` clause. Even though we handled all the values in the Vanilla enum, and our own, this would throw if another mod adds a different entry.

How can you prevent this? There is no single universal way to avoid crashes - your approach should be adapted on a case-by-case basis. In general, though:

> - If the `switch` expression is in a vanilla method, you can use a mixin to edit it

Reviewer 3: wouldn't it be cool if there was any elaboration on how to edit such a switch? i think just a hint in the right direction should be good enough, e.g. whats the best practice for the injector and target (at least for the situation in the code above)

Reviewer 2: No universal advice exists, at best people can be pointed to the discord / the general mixin injector pages

---
title: Enum Extension
description: Learn how to add entries to enums with Mixin and Class Tweakers.
authors:
  - MildestToucan
---

Enum extension is a Mixin feature that can reliably add new entries to an enum.

When targeting Minecraft enums, the mixin can be combined with [class tweaking](../class-tweakers) to display the new enum entries

> to the decompiled source. This can be [transitive](../class-tweakers/index#transitive-entries) to allow mods depending on yours to see your added entries.

Reviewer: ```suggestion
When targeting Minecraft enums, you can use mixins together with [class tweaking](../class-tweakers) to append new enum entries
to the decompiled source. If that is set to be [transitive](../class-tweakers/index#transitive-entries), mods that depend on yours will also see your added entries.
```

alternatively, if "display" is the intended verb in this context:

```suggestion
When targeting Minecraft enums, you can use mixins together with [class tweaking](../class-tweakers) to display new enum entries
in the decompiled source. If that is set to be [transitive](../class-tweakers/index#transitive-entries), mods that depend on yours will also see your added entries.
```

::: warning

Enum extension requires Loader 0.19.0 for Mixin support and Loom 1.16 for class tweaker support.

> Additionally, class tweaker file headers must specify a version of `v2` to use enum extensions.

Reviewer: ```suggestion
Additionally, class tweaker file headers must specify `v2` as the version to use enum extensions.
```

:::

## Creating the Mixin {#creating-the-mixin}

Before creating the mixin class, make sure that you are depending explicitly on Loader 0.19.0 or above

> in your `fabric.mod.json` file:

Reviewer: ```suggestion
Before creating the mixin class, make sure that Loader 0.19.0 or above is an explicit dependency
in your `fabric.mod.json` file:
```

part of the targeted enum class. For example, adding a new `RecipeBookType` entry would look like this:

<<< @/reference/latest/src/main/java/com/example/docs/mixin/class_tweakers/RecipeBookTypeMixin.java#enum-extension-no-impls-example-mixin

Since there are no fields or methods to implement in the target enum, there's no need to do anything else.

::: warning

Added enum constants should always be prefixed by your mod id to ensure uniqueness. To preserve capitalization, the prefix would typically be
`MODID_`.

:::

If the enum entries need fields or to implement methods, you should shadow the relevant constructor by creating a matching one and annotating it with
`@Shadow`. For example, adding a new `RecipeCategory` entry would look like this:

<<< @/reference/latest/src/main/java/com/example/docs/mixin/class_tweakers/RecipeCategoryMixin.java#enum-extension-ctor-impls-example-mixin

## Making the Class Tweaker Entry {#making-the-class-tweaker-entry}

If you are targeting a Minecraft enum, you can use a class tweaker entry to make the added entry appear in the target enum.

> To opt into this feature, remember to use Loom 1.16 or above, and to have the file [header](../class-tweakers/index#file-format) have its

Reviewer 4: Have its...?

Author: Oh boy I really need to double check myself. Seems like I was a bit hasty with some phrasing lmao. I'll have that fixed soon.

---
title: Enum Extension
description: Learn how to add entries to enums with Mixin and Class Tweakers.
authors:
  - cassiancc
  - MildestToucan
---

Enum extension is a Mixin feature that can reliably add new entries to an enum.

When targeting Minecraft enums, the mixin can be combined with [class tweaking](../class-tweakers) to display the new enum entries
to the decompiled source. This can be [transitive](../class-tweakers/index#transitive-entries) to allow mods depending on yours to see your added entries.

::: warning

> Enum extension requires Loader 0.19.0 for Mixin support and Loom 1.16 for class tweaker support.

Reviewer 4: ```suggestion
Enum extension requires at least Loader 0.19.0 for Mixin support and at least Loom 1.16 for class tweaker support.
```

Additionally, class tweaker file headers must specify a version of `v2` to use enum extensions.

:::

## Creating the Mixin {#creating-the-mixin}

Before creating the mixin class, make sure that you are depending explicitly on Loader 0.19.0 or above
in your `fabric.mod.json` file:

```json:no-line-numbers
...
"depends": {
  ...
  "fabricloader": ">=0.19.0"
  ...
}

> ...

Reviewer: did you know that the plurals of Ellipsis and of Ellipse are both Ellipses? This is a case of [polysemy](https://en.wikipedia.org/wiki/Polysemy), which is when a word has multiple different meanings that have a common origin. The common origin is the Greek [ἔλλειψις](https://en.wiktionary.org/wiki/%E1%BC%94%CE%BB%CE%BB%CE%B5%CE%B9%CF%88%CE%B9%CF%82), which is also the origin of the word Eclipse (the kind of eclipse that the Sun does every few years in bowing respect of the ominous IDE. Obviously, I'm referring to Android Studio, which when run on my laptop could easily replace the sun in providing heat for the Earth).

Polysemes are not to be confused with [Homonyms](https://en.wikipedia.org/wiki/Homonym), which happen coincidentally. An example of the latter is "left" (past of leave) and "left" (political orientation). Instead, polysemes can be traced to a common origin.

However, be careful! According to most sources I could find online, "Ellipses" is pronounced differently depending on the meaning:
- if it's plural for ellipsis (which is read eh-llip-sis): eh-llip-sees
- if it's plural for ellipse (which is read eh-llips): eh-llip-sis

That means, apparently, that ellipsis and ellipses (the shapes) are [homophones](https://en.wikipedia.org/wiki/Homophone), also known as heterographs. Although Wikipedia says homophones are a type of homonymy, which we said does not apply in this case, so idrk

this isn't really a review comment, i'm realizing just now. welp, I guess it's a freebie that you can mark as resolved!

Author: My quota is gonna benefit so much from this. Thanks Miroma!

```

> Even if you are using the correct Loader version as a Gradle dependency, you must explicitly depend on it in

Reviewer 4: ```suggestion
Even if you are using the correct Loader version as a Gradle dependency, you must explicitly depend on at least version 0.19.0 in
```

> order to opt into new Mixin features.

Reviewer: ```suggestion
order to opt into this Mixin feature.
```

shiny new isn't new forever: https://en.wikipedia.org/wiki/Pont_Neuf
> The **Pont Neuf** (French for "New Bridge") is the oldest standing bridge across the river Seine in Paris, France.

Reviewer 4: Maybe "the features introduced in this version"?

Reviewer: This is the Enum Extension article specifically, that's why I restricted it to the singular

Author: Yeah Miroma's suggestion makes the most sense to me here. We could add something about the loader dep being related to Mixin feature opt-in on the FMJ page at some point, if that's not already planned.

To make an enum extension, create a `@Mixin`-annotated enum class, and add your constants to it as if they were

> part of the targeted enum class. For example, adding a new `RecipeBookType` entry would look like this:

Reviewer: ```suggestion
To make an enum extension, create an `enum`, annotate it with `@Mixin`, and add your constants to it as if they were
part of the targeted enum class. For example, let's add a new entry to `RecipeBookType`:
```

<<< @/reference/latest/src/main/java/com/example/docs/mixin/class_tweakers/RecipeBookTypeMixin.java#enum-extension-no-impls-example-mixin

Since there are no fields or methods to implement in the target enum, there's no need to do anything else.

::: warning

Added enum constants should always be prefixed by your mod id to ensure uniqueness. To preserve capitalization, the prefix would typically be
`MODID_`.

:::

If the enum entries need fields or to implement methods, you should shadow the relevant constructor by creating a matching one and annotating it with
`@Shadow`. For example, adding a new `RecipeCategory` entry would look like this:

> <<< @/reference/latest/src/main/java/com/example/docs/mixin/class_tweakers/RecipeCategoryMixin.java#enum-extension-ctor-impls-example-mixin

Reviewer: ```suggestion
::: warning IMPORTANT

You should always prefix the enum constants you add with your mod ID to ensure uniqueness. For these Docs, we'll use `EXAMPLE_MOD_`.

:::

### Additional Fields and Methods {#additional-fields-and-methods}

If the targeted enum declares additional fields or method implementations, you'll have to shadow the relevant constructor.

For example, let's add a new `RecipeCategory` entry. Create a constructor matching the target, and annotate it with `@Shadow`.

<<< @/reference/latest/src/main/java/com/example/docs/mixin/class_tweakers/RecipeCategoryMixin.java#enum-extension-ctor-impls-example-mixin
```

imo it reads better like this, without fore@Shadowing (pun intended) the next section

## Making the Class Tweaker Entry {#making-the-class-tweaker-entry}

> If you are targeting a Minecraft enum, you can use a class tweaker entry to make the added entry appear in the target enum.

Reviewer: ```suggestion
If you are targeting a Minecraft enum, you can use a class tweaker entry to actually modify the target enum.
```

Author: I'm not sure about this one. The enum *is* being "actually" modified by the Mixin at runtime. 

Class tweaking only takes care of making it appear in the decompiled source so it can be referenced directly and be seen by your dependents.

The suggested phrasing here imo is misleading as it implies that without class tweaking, you're not "actually" modifying the enum.

Reviewer: "to visibly modify the target enum [in the decompiled source]"?

Author: Yeah, "to visibly modify the target enum in the decompiled source" sounds good to me.

> To opt into this feature, remember to use Loom 1.16 or above, and to have the file [header](../class-tweakers/index#file-format) version as `v2`.

Reviewer: ```suggestion
To opt into this feature, remember to use Loom 1.16 or above, and to set the [file header version](../class-tweakers/index#file-format) to `v2`.
```

The syntax for an enum extension entry is:

```classtweaker:no-line-numbers
extend-enum    <targetClassName>    <ENUM_CONSTANT_NAME>
```

For class tweaking, classes use their [internal names](../mixins/bytecode#class-names).

For example, the class tweaker entries for the constants we added in the [mixin section](#creating-the-mixin) would respectively be:

<<< @/reference/latest/src/main/resources/example-mod.classtweaker#enum-extension-no-impls-example-entry

and

<<< @/reference/latest/src/main/resources/example-mod.classtweaker#enum-extension-ctor-impls-example-entry

## Applying Changes {#applying-changes}

To see your added enum entries in the decompiled source, you must refresh your Gradle project by [regenerating sources](../getting-started/generating-sources).
If modifications do not appear, you can try [validating](../class-tweakers/index#validating-the-file) the file and checking if any errors appear.

The added entries should now appear in the decompiled source. If the enum has fields or other elements for its entries to implement, those will not appear for your
added entry, and are instead left to your mixin to implement at runtime.

> You can now use the enum constant from your code:

Reviewer: ```suggestion
You'll have to refresh your Gradle project by [regenerating sources](../getting-started/generating-sources) before you can see your added enum entries in the decompiled source.
If modifications do not appear, you can try [validating](../class-tweakers/index#validating-the-file) the file and checking if any errors appear.

::: info

You will not see [fields, methods or other elements](#additional-fields-and-methods) in the decompiled source code. That's because those are handled by the mixin, and are only applied at runtime.

:::

You can now use the enum constant in your code:
```

---
title: Enum Extension
description: Learn how to add entries to enums with Mixin and Class Tweakers.
authors:
  - cassiancc
  - MildestToucan
---

Enum extension is a Mixin feature that can reliably add new entries to an enum.

When targeting Minecraft enums, you can use mixins together with [class tweaking](../class-tweakers) to display new enum entries
in the decompiled source. If that is set to be [transitive](../class-tweakers/index#transitive-entries), mods that depend on yours will also see your added entries.

::: warning

Enum extension requires at least Loader 0.19.0 for Mixin support and at least Loom 1.16 for class tweaker support.

Additionally, class tweaker file headers must specify `v2` as the version to use enum extensions.

:::

## Creating the Mixin {#creating-the-mixin}

Before creating the mixin class, make sure that Loader 0.19.0 or above is an explicit dependency
in your `fabric.mod.json` file:

```json:no-line-numbers
...
"depends": {
  ...
  "fabricloader": ">=0.19.0"
  ...
}
...
```

Even if you are using the correct Loader version as a Gradle dependency, you must explicitly depend on at least version 0.19.0 in
order to opt into this Mixin feature.

> To make an enum extension, create an `enum`, annotate it with `@Mixin`, and add your constants to it as if they were

Reviewer 4: ```suggestion
To make an enum extension, create an `enum` in your `mixin` package, annotate it with `@Mixin`, and add your constants to it as if they were
```

Reviewer 5: imo "mixin" is better than "`mixin`"

Author: Will implement this without the monospacing as Jab125 is suggesting. Technically the mixin package could be named other things, so monospace might be a little misleading.
